fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerRegistryImpl.java

```diff
@@ -0,0 +1,130 @@
+ * loader to determine which {@link LuceneAnalyzerFactory} implementation exist,
+ * and it populates the registry with those analyzers. An instance of this registry
+ * is used by the {@link LuceneIndexMaintainer}
+ * in order to choose the analyzer for a block of text. One can therefore register
+ * additional analyzers for that index by calling the {@link #register(LuceneAnalyzerFactory) register}
+ * method on the singleton instance of this class and supplying the additional analyzer.
+ */
+public class LuceneAnalyzerRegistryImpl implements LuceneAnalyzerRegistry {
+    @Nonnull
+    private static final Logger LOGGER = LoggerFactory.getLogger(LuceneAnalyzerRegistryImpl.class);
+    @Nonnull
+    private static final LuceneAnalyzerRegistryImpl INSTANCE = new LuceneAnalyzerRegistryImpl();
+
+    @Nonnull
+    private Map<String, LuceneAnalyzerFactory> registry;
+
+    @Nonnull
+    private static Map<String, LuceneAnalyzerFactory> initRegistry() {
+        final Map<String, LuceneAnalyzerFactory> registry = new HashMap<>();
+        for (LuceneAnalyzerFactory factory : ServiceLoader.load(LuceneAnalyzerFactory.class)) {
+            final String name = factory.getName();
+            if (registry.containsKey(name)) {
+                if (LOGGER.isWarnEnabled()) {
+                    LOGGER.warn(KeyValueLogMessage.of("duplicate lucene analyzer", LogMessageKeys.ANALYZER_NAME, name));
+                }
+            } else {
+                if (LOGGER.isInfoEnabled()) {
+                    LOGGER.info(KeyValueLogMessage.of("found lucene analyzer", LogMessageKeys.ANALYZER_NAME, name));
+                }
+                registry.put(name, factory);
+            }
+        }
+        return registry;
+    }
+
+    @Nonnull
+    public static LuceneAnalyzerRegistry instance() {
+        return INSTANCE;
+    }
+
+    private LuceneAnalyzerRegistryImpl() {
+        registry = initRegistry();
+    }
+
+    @Nonnull
+    @Override
+    public Map<String, LuceneAnalyzerFactory> getRegistry() {
+        return Collections.unmodifiableMap(registry);
+    }
+
+    @Nonnull
+    @Override
+    public Analyzer getLuceneAnalyzer(@Nonnull Index index) {
+        final String name = index.getOption(IndexOptions.TEXT_ANALYZER_NAME_OPTION);
+        // TODO: Get rid of the condition after OR operator, after having all analyzers registered with this registry
+        if (name == null || !name.equals(NgramAnalyzer.getName())) {
+            return new StandardAnalyzer();
+        } else {
+            LuceneAnalyzerFactory analyzerFactory = registry.get(name);
+            if (analyzerFactory == null) {
+                throw new MetaDataException("unrecognized lucene analyzer for tokenizer", LogMessageKeys.ANALYZER_NAME, name);
+            }
+            return analyzerFactory.getAnalyzer(index);
+        }
+    }
+
+    // Synchronize this method so that we don't need a ConcurrentHashMap but so that
+    // it is still thread safe.
```

Comment:
What prevents this method growing/rehashing the hash map while someone is trying to access the result of `getRegistry`?

Reply:
I just used same abstraction here as ```TextTokenizerRegistry```. In fact I am not really sure in what cases ```getRegistry()``` could be needed. I am guessing probably ```getRegistry()``` and ```register``` are not supposed to be called concurrently? @alecgrieser do you have any ideas about the design of ```TextTokenizerRegistry```? If so, any thoughts about the potential use cases of ```getRegistry()``` and the dynamic ```register```?

Reply:
It's been a bit since I've looked at this code, so my understanding is a bit rusty, but I believe the intention here was primarily for `register` to be used during tests (and that's, in fact, the only place I can find it being used). That way test-only tokenizers could be created and swapped in during tests with different properties to test, for example, certain error scenarios. However, it also offers an alternative for people who don't want to set up the `ServiceLoader` correctly, I guess, but that should really only be done in some kind of static initialization block, not while someone is trying to access the registry. 

It does seem like there's a concurrency bug here (i.e., in both registries)  if `register` is used outside of the above contexts. If you don't need to use `register` (and can instead rely on only using the `ServiceLoader`), then that's one way of getting around the problem. Otherwise, we either need to document or fix the concurrency bug, I think

Reply:
@alecgrieser thank you! I just removed the ```register```, ```getRegistry``` and ```reset``` to get rid of the support for registration in fly, considering it is not actively used for now. Please let me know this looks good to you. @MMcM @alecgrieser